kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Coordinator.java

```diff
@@ -187,29 +194,72 @@ private void commitToTable(
 
     Map<Integer, Long> committedOffsets = lastCommittedOffsetsForTable(table, branch.orElse(null));
 
-    List<CommitResponsePayload> payloads =
-        envelopeList.stream()
+    List<Envelope> payloads =
+        deduplicateEnvelopes(envelopeList, commitState.currentCommitId(), tableIdentifier).stream()
             .filter(
                 envelope -> {
                   Long minOffset = committedOffsets.get(envelope.partition());
                   return minOffset == null || envelope.offset() >= minOffset;
                 })
-            .map(envelope -> (CommitResponsePayload) envelope.event().payload())
             .collect(toList());
 
     List<DataFile> dataFiles =
-        payloads.stream()
-            .filter(payload -> payload.dataFiles() != null)
-            .flatMap(payload -> payload.dataFiles().stream())
-            .filter(dataFile -> dataFile.recordCount() > 0)
-            .collect(toList());
+        deduplicateBatch(
+            payloads.stream()
+                .flatMap(
+                    envelope -> {
+                      CommitResponsePayload payload =
+                          (CommitResponsePayload) envelope.event().payload();
```

Comment:
Unfortunate this isn't typed.  Big assumption this is always going to be a safe cast, but I've convinced myself it's ok for now.

Reply:
Yep, the code was doing that before as well.